### PR TITLE
Organize test specs with ginkgo

### LINKED ISSUE #995: e2e: organize test cases with a testing framework
## Feature Request

**Is your feature request related to a problem? Please describe:**

- unable to test part of test cases
- no unified test report

**Describe the feature you'd like:**

organize test cases with a testing framework, e.g.

- go: https://golang.org/pkg/testing/
- ginkgo: https://github.com/onsi/ginkgo

I prefer ginkgo, because it's used by Kubernetes and we can reuse a lot of code in Kubernetes testing framework: https://github.com/kubernetes/kubernetes/tree/v1.16.0/test/e2e/framework.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by cofyc**
### What problem does this PR solve? 

fixes #995

### Check List 

Tests 

 - Unit test
 - E2E test

### CODE REVIEW

File: tests/pkg/metrics/annotation_util.go [hunk 1]
```diff
@@ -130,7 +130,7 @@ func (cli *Client) getAnnotationPath() string {
 	return u.String()
 }
 
-func init() {
+func StartServer() {
```
Comment (cofyc) [inline tests/pkg/metrics/annotation_util.go]: don't start by default

File: tests/pkg/client/client.go [hunk 2]
```diff
@@ -23,7 +23,7 @@ var (
 	kubeconfigPath string
 )
 
-func init() {
+func RegisterFlags() {
```
Comment (cofyc) [inline tests/pkg/client/client.go]: run it for stability only

File: hack/run-e2e.sh [hunk 3]
```diff
@@ -10,6 +10,7 @@ cd $ROOT
 source $ROOT/hack/lib.sh
 
 hack::ensure_kubectl
+hack::ensure_helm
```
Comment (cofyc) [inline hack/run-e2e.sh]: same helm version in e2e image

File: tests/e2e/e2e.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+// This is modified from framework.SetupSuite().
+// setupSuite is the boilerplate that can be used to setup ginkgo test suites, on the SynchronizedBeforeSuite step.
+// There are certain operations we only want to run once per overall test invocation
+// (such as deleting old namespaces, or verifying that all system pods are running.
+// Because of the way Ginkgo runs tests in parallel, we must use SynchronizedBeforeSuite
+// to ensure that these operations only run on the first parallel Ginkgo node.
+//
+// This function takes two parameters: one function which runs on only the first Ginkgo node,
+// returning an opaque byte array, and then a second function which runs on all Ginkgo nodes,
+// accepting the byte array.
```
Comment (aylei) [inline tests/e2e/e2e.go]: These comments are not coordinated with the function signature

**Comment by cofyc** [comment]
sure, `./hack/e2e.sh -h` will show some basic examples

**Comment by weekface** [comment]
please resolve the conflicts.

**Comment by cofyc** [comment]
/run-e2e-test
resoved, PTAL

**Comment by cofyc** [comment]
@aylei @weekface @tennix @onlymellb take a look?

File: tests/images/e2e/Dockerfile [hunk 5]
```diff
@@ -19,4 +19,5 @@ ADD tidb-cluster /charts/e2e/tidb-cluster
 ADD tidb-backup /charts/e2e/tidb-backup
 ADD manifests /manifests/e2e
 
-ADD bin/e2e /usr/local/bin/e2e
+ADD bin/ginkgo /usr/local/bin/
```
Comment (tennix) [inline tests/images/e2e/Dockerfile]: Where does this binary come from?

Comment (cofyc) [reply tests/images/e2e/Dockerfile]: built by `make e2e-build`

Comment (tennix) [reply tests/images/e2e/Dockerfile]: Oh, I see. But do we need a specific ginkgo version? When enabling GO111MODULE, the go.sum will be changed if we always pull the latest ginkgo.

Comment (cofyc) [reply tests/images/e2e/Dockerfile]: I locked the ginkgo version (see go.mod changes). `go build <full-path-library>` in a go modules project will build the version in `go.mod` file. (will try to fetch the latest only if the version is missing)

**Pull request merged**